contracts/RandomBeaconHistory.cdc

```diff
@@ -0,0 +1,119 @@
+    //
+    /// A publicly queryable interface for the Heartbeat resource.
+    ///
+    pub resource interface HeartbeatPublic {
+        pub fun getRandomSourceHistory(): [[UInt8]]
+        pub fun getInitHeight(): UInt64
+        pub fun sourceOfRandomness(atBlockHeight blockHeight: UInt64): [UInt8]
+    }
+
+    /// The Heartbeat resource containing each block's source of randomness in sequence
+    ///
+    pub resource Heartbeat : HeartbeatPublic {
+        /// The height at which the first source of randomness was recorded
+        access(self) var initHeight: UInt64?
+        /// Sequence of random sources recorded by the Heartbeat, stored as an array over a mapping to reduce storage
+        // TODO: Test at scale
+        access(self) let randomSourceHistory: [[UInt8]]
+
+        init() {
+            self.initHeight = nil
+            self.randomSourceHistory = []
+        }
+
+        /// Getter for the totality of recorded randomness source history
+        ///
+        /// @return An array of random sources, each source an array of UInt8
+        ///
+        pub fun getRandomSourceHistory(): [[UInt8]] {
+            pre {
+                self.initHeight != nil: "Heartbeat has not been initialized"
+            }
+            return self.randomSourceHistory
+        }
+
+        /// Getter for the block height at which the first source of randomness was recorded
+        ///
+        /// @return The block height at which the first source of randomness was recorded
+        ///
+        pub fun getInitHeight(): UInt64 {
+            return self.initHeight ?? panic("Heartbeat has not been initialized")
+        }
+
+        /// Getter for the source of randomness at a given block height. Panics if the requested block height either
+        /// precedes or exceeds the recorded history.
+        ///
+        /// @param atBlockHeight The block height at which to retrieve the source of randomness
+        /// @return The source of randomness at the given block height
+        ///
+        pub fun sourceOfRandomness(atBlockHeight: UInt64): [UInt8] {
+            pre {
+                self.initHeight != nil: "Heartbeat has not been initialized"
+                atBlockHeight >= self.initHeight!: "Requested block height precedes recorded history"
+                atBlockHeight < getCurrentBlock().height: "Source of randomness not yet recorded"
+            }
+            let index = UInt64(atBlockHeight - self.initHeight!)
+
+            assert(
+                index >= 0 && index < UInt64(self.randomSourceHistory.length),
+                message: "Problem finding random source history index"
+            )
+
+            return self.randomSourceHistory[index]
+        }
+
+        /// Callable by owner of the Heartbeat resource, Flow Service Account, records the provided random source
+        ///
+        /// @param randomSourceHistory The random source to record
+        ///
+        pub fun heartbeat(randomSourceHistory: [UInt8]) {
```

Comment:
IMO the heartbeat resource should have only one method `pub fun heartbeat(randomSourceHistory: [UInt8])` the rest should be on the contract itself.